File: 1846-maximum-element-after-decreasing-and-rearranging/1846-maximum-element-after-decreasing-and-rearranging.py
class Solution:
    def maximumElementAfterDecrementingAndRearranging(self, arr: List[int]) -> int:
        n = len(arr)
        counts = [0] * (n + 1)
        
        for num in arr:
            counts[min(num, n)] += 1
        
        ans = 1
        for num in range(2, n + 1):
            ans = min(ans + counts[num], num)
        
        return ans

File: 1846-maximum-element-after-decreasing-and-rearranging/1846-maximum-element-after-decreasing-and-rearranging.cpp
class Solution {
public:
    int maximumElementAfterDecrementingAndRearranging(vector<int>& arr) {
        sort(arr.begin(), arr.end());
        if (arr[0] != -1)
            arr[0] = 1;
        
        int ans = 1;
        
        for (int i = 0; i < arr.size() - 1; ++i){
            if (abs(arr[i + 1] - arr[i]) > 1)
                arr[i + 1] = arr[i] + 1;
            ans = max(ans, max(arr[i + 1], arr[i]));
        }
        
        return ans;
    }
};